File: clients/windows/Mailcal/Services/EditorWebView.cs
// The hardening every host of the shared editor bundle applies, the message composer AND the
// Settings signature editor. It lives here, in ONE place, because it is a security contract rather
// than a per-screen detail: authoring a signature is authoring mail content, so it gets the
// composer's gates, not a lighter set (docs/composer-security.md, "Layer 3, Native WebView host
// gates"; docs/signatures.md). Two hosts with two copies of these settings is two chances for one of
// them to drift, which is exactly why Android collapsed its two into EditorWebView.kt.
//
// What it guarantees: script runs for the bundled local document only, no host objects, no web
// messages, no default context menu, no new windows, every navigation away is cancelled, and every
// http/https subresource request is answered 403, the native barrier behind the bundle's own CSP.
//
// Do not relax one of these without updating that doc (rule AND matrix) and every other platform.

using System.Text.Json;
using Microsoft.UI.Xaml.Controls;
using Microsoft.Web.WebView2.Core;

namespace Allodia.Mailcal.Services;

/// <summary>
/// One hardened host of <c>clients/composer/dist/editor.html</c>. Owns the WebView2's gate wiring and
/// the single navigation the host is allowed to perform, so a caller only ever loads the bundle and
/// talks to it through <c>window.*</c> hooks.
/// </summary>
internal sealed class EditorWebViewHost
{
    private readonly WebView2 _view;
    private Task? _init;

    // The one navigation this host permits: our own NavigateToString of the bundle. Everything
    // after it is cancelled, so a link, a redirect, or a script-driven location change cannot take
    // the editor anywhere.
    private bool _expectingLoad;

    /// <summary>Wraps <paramref name="view"/>; the gates are applied on the first load.</summary>
    internal EditorWebViewHost(WebView2 view) => _view = view;

    /// <summary>
    /// Invoked once the bundle has finished loading, at which point its <c>window.*</c> hooks are
    /// defined. Every seed a host injects belongs here and nowhere earlier: a hook called before the
    /// page has parsed lands on an undefined function and fails silently, leaving the editor in its
    /// default state with no error anywhere.
    /// </summary>
    internal Func<Task>? PageReady { get; set; }

    /// <summary>The underlying WebView2 environment, or <c>null</c> before the first load. Callers
    /// use it to tell "the page is up" from "nothing has loaded yet".</summary>
    internal CoreWebView2? Core => _view.CoreWebView2;

    /// <summary>Applies the gates (once) and loads the shared editor bundle.</summary>
    internal async Task LoadAsync()
    {
        await EnsureAsync();
        _expectingLoad = true;
        _view.CoreWebView2!.NavigateToString(Bundle());
    }

    /// <summary>Runs <paramref name="script"/> in the editor document, discarding its result.</summary>
    internal async Task RunAsync(string script)
    {
        await EnsureAsync();
        await _view.CoreWebView2!.ExecuteScriptAsync(script);
    }

    /// <summary>Runs <paramref name="script"/> and decodes its result as a JavaScript string,
    /// what every hook that hands data back (<c>composerDocument()</c>, <c>signatureBody()</c>)
    /// returns.</summary>
    internal async Task<string?> ReadStringAsync(string script)
    {
        await EnsureAsync();
        var encoded = await _view.CoreWebView2!.ExecuteScriptAsync(script);
        return JsonSerializer.Deserialize<string>(encoded);
    }

    /// <summary>Releases the WebView2 backing this host. Safe to call on one that never
    /// initialised.</summary>
    internal void Close()
    {
        try
        {
            _view.Close();
        }
        catch (Exception ex)
        {
            // A host torn down before its WebView2 ever initialised has nothing to close.
            Log.Warn($"editor: teardown ({ex.GetType().Name})");
        }
    }

    /// <summary>Encodes <paramref name="value"/> as a JavaScript string literal, so host-supplied
    /// text can be passed into a hook without breaking out of the argument.</summary>
    internal static string Arg(string value) => JsonSerializer.Serialize(value);

    private Task EnsureAsync() => _init ??= InitAsync();

    private async Task InitAsync()
    {
        await _view.EnsureCoreWebView2Async();
        var core = _view.CoreWebView2;
        var settings = core.Settings;
        // Script runs ONLY for the local editor document; neither host ever renders untrusted mail
        // in this WebView2 (the reading pane keeps its own, with scripting off).
        settings.IsScriptEnabled = true;
        settings.AreHostObjectsAllowed = false;
        settings.IsWebMessageEnabled = false;
        settings.AreDefaultContextMenusEnabled = false;
        core.NavigationStarting += OnNavigationStarting;
        core.NavigationCompleted += OnNavigationCompleted;
        core.NewWindowRequested += (_, args) => args.Handled = true;
        core.AddWebResourceRequestedFilter("*", CoreWebView2WebResourceContext.All);
        core.WebResourceRequested += OnWebResourceRequested;
    }

    private void OnNavigationStarting(CoreWebView2 sender, CoreWebView2NavigationStartingEventArgs args)
    {
        if (_expectingLoad)
        {
            _expectingLoad = false;
            return;
        }
        args.Cancel = true;
    }

    private async void OnNavigationCompleted(CoreWebView2 sender, CoreWebView2NavigationCompletedEventArgs args)
    {
        if (!args.IsSuccess || PageReady is not { } ready)
        {
            return;
        }
        try
        {
            await ready();
        }
        catch (Exception ex)
        {
            // The host's own seeding failed. Never let it escape an async void handler, which would
            // take the process down; the host reports what it could not do.
            Log.Warn($"editor: page-ready seeding ({ex.GetType().Name})");
        }
    }

    private void OnWebResourceRequested(CoreWebView2 sender, CoreWebView2WebResourceRequestedEventArgs args)
    {
        var uri = args.Request.Uri ?? string.Empty;
        if (uri.StartsWith("http://", StringComparison.OrdinalIgnoreCase)
            || uri.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
        {
            args.Response = sender.Environment.CreateWebResourceResponse(null, 403, "Blocked", string.Empty);
        }
    }

    // The shared bundle, copied next to the exe by the csproj. The fallback defines both hosts'
    // read-back hooks so a missing asset degrades to an empty document rather than an exception at
    // the first ExecuteScriptAsync.
    private static string Bundle()
    {
        var asset = Path.Combine(AppContext.BaseDirectory, "composer", "editor.html");
        if (File.Exists(asset))
        {
            return File.ReadAllText(asset);
        }
        return "<!doctype html><html><body><script>"
            + "window.composerDocument=function(){return JSON.stringify({blocks:[],attachments:[]});};"
            + "window.signatureBody=function(){return JSON.stringify({body_html:\"\",body_plain:\"\"});};"
            + "</script></body></html>";
    }
}

File: clients/android/app/src/main/java/eu/allodia/mailcal/EditorWebView.kt
// The hardening every host of the shared editor bundle applies (docs/composer-security.md): the
// message composer AND the Settings signature editor. It lives here, in one place, because it is a
// security contract rather than a per-screen detail, authoring a signature is authoring mail
// content, so it gets the composer's gates, not a lighter set. Two hosts with two copies of these
// settings is two chances for one of them to drift.
//
// What it guarantees: JavaScript runs for the bundled document only, no window can be opened, no
// file or content:// URL is readable, no DOM storage persists, no mixed content loads, every
// navigation away is refused, and every http/https subresource request is answered with an empty
// body, the native barrier behind the bundle's own CSP.
package eu.allodia.mailcal

import android.content.ClipData
import android.content.ClipboardManager
import android.webkit.WebResourceRequest
import android.webkit.WebResourceResponse
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.PopupMenu
import java.io.ByteArrayInputStream

internal fun WebView.applyEditorSecuritySettings() {
    settings.javaScriptEnabled = true
    settings.javaScriptCanOpenWindowsAutomatically = false
    settings.setSupportMultipleWindows(false)
    settings.allowFileAccess = false
    settings.allowContentAccess = false
    settings.domStorageEnabled = false
    settings.mixedContentMode = WebSettings.MIXED_CONTENT_NEVER_ALLOW
}

// Long-pressing a link offers its address.
//
// Selected text already gets Cut/Copy/Paste from the system's own selection bar, so that half needs
// nothing; a link gets nothing by default, and a link inside a quoted original cannot be opened by
// tapping it either (navigation is blocked), so without this it is text the user can see and not
// use. The same item every other client offers (docs/composer-security.md, Gate 14).
//
// Returning false for anything that is not a link leaves ordinary text selection alone.
internal fun WebView.installEditorLinkMenu() {
    setOnLongClickListener { view ->
        val result = (view as WebView).hitTestResult
        val url = result.extra
        val isLink = result.type == WebView.HitTestResult.SRC_ANCHOR_TYPE ||
            result.type == WebView.HitTestResult.SRC_IMAGE_ANCHOR_TYPE
        if (!isLink || url.isNullOrBlank()) {
            return@setOnLongClickListener false
        }
        PopupMenu(context, view).apply {
            menu.add(L10n.action_copy_link(context)).setOnMenuItemClickListener {
                context.getSystemService(ClipboardManager::class.java)
                    ?.setPrimaryClip(ClipData.newPlainText(null, url))
                true
            }
        }.show()
        true
    }
}

// The navigation + subresource gate. Subclasses add their own `onPageFinished` (the hooks each host
// injects once the bundle has parsed) and inherit both refusals.
internal open class EditorWebViewClient : WebViewClient() {
    override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean = true

    override fun shouldInterceptRequest(
        view: WebView?,
        request: WebResourceRequest?,
    ): WebResourceResponse? {
        val scheme = request?.url?.scheme?.lowercase()
        return if (scheme == "http" || scheme == "https") {
            WebResourceResponse("text/plain", "utf-8", ByteArrayInputStream(ByteArray(0)))
        } else {
            null
        }
    }
}
